src/sage/sets/primes.py

```diff
@@ -1,187 +1,851 @@
+            sage: 13 in P
+            True
+
+        TESTS::
+
+            sage: x in P
             False
         """
         try:
             if x not in ZZ:
                 return False
-            return ZZ(x).is_prime()
         except TypeError:
             return False
+        x = ZZ(x)
+        if not x.is_prime():
+            return False
+        e = self._exceptions.get(x, None)
+        return (e is True) or (e is None and x % self._modulus in self._classes)
 
-    def _an_element_(self):
+    def cardinality(self):
+        r"""
+        Return the cardinality of this set.
+
+        EXAMPLES::
+
+            sage: P = Primes(modulus=4); P
+            Set of prime numbers congruent to 1 modulo 4: 5, 13, 17, 29, ...
+            sage: P.cardinality()
+            +Infinity
+
+        ::
+
+            sage: P = Primes(modulus=4, classes=[2]); P
+            Finite set of prime numbers: 2
+            sage: P.cardinality()
+            1
         """
-        Return a typical prime number.
+        if self.is_finite():
+            return ZZ(len(self._elements))
+        return infinity
+
+    def first(self, n=None):
+        r"""
+        Return the first element in this set.
+
+        EXAMPLES::
+
+            sage: P = Primes(modulus=5); P
+            Set of prime numbers congruent to 1 modulo 5: 11, 31, 41, 61, ...
+            sage: P.first()
+            11
+        """
+        if self.is_empty():
+            raise EmptySetError
+        return self.next(1)
+
+    def next(self, x):
+        r"""
+        Return the smallest element in this set strictly
+        greater than ``x``.
+
+        INPUT:
+
+        - ``x`` -- an integer
+
+        EXAMPLES::
+
+            sage: P = Primes(modulus=5); P
+            Set of prime numbers congruent to 1 modulo 5: 11, 31, 41, 61, ...
+            sage: P.next(1000)
+            1021
+
+        If there is no element greater than the given bound, an
+        error is raised::
+
+            sage: P = Primes(modulus=0, classes=[2, 5]); P
+            Finite set of prime numbers: 2, 5
+            sage: P.next(10)
+            Traceback (most recent call last):
+            ...
+            ValueError: no element greater that 10 in this set
+        """
+        x = ZZ(x)
+        if not self._classes:
+            if not (self._elements and x < self._elements[-1]):
+                raise ValueError("no element greater that %s in this set" % x)
+            min = 0
+            max = len(self._elements)
+            while min < max:
+                i = (min + max) // 2
+                if self._elements[i] <= x:
+                    min = i + 1
+                if self._elements[i] > x:
+                    max = i
+            return self._elements[min]
+        while True:
+            x = x.next_prime()
+            e = self._exceptions.get(x, None)
+            if (e is True) or (e is None and x % self._modulus in self._classes):
+                return x
```

Comment:
Not sure whether the following suggestion is any easier to read, feel free to ignore:
```suggestion
        if self._classes:
            while True:
                x = x.next_prime()
                e = self._exceptions.get(x, None)
                if (e is True) or (e is None and x % self._modulus in self._classes):
                    return x

        if not self._elements or x >= self._elements[-1]:
            raise ValueError("no element greater that %s in this set" % x)
        min = 0
        max = len(self._elements)
        while min < max:
            i = (min + max) // 2
            if self._elements[i] <= x:
                min = i + 1
            if self._elements[i] > x:
                max = i
        return self._elements[min]
```